third_party/nvfuser/csrc/scheduler/normalization.cpp

```diff
@@ -1,35 +1,232 @@
+  // threads_per_block_max
+  int64_t threads_per_block_max = inner_dim_numel >= 20480 ? 512 : 256;
+  while (batch > batch_min && threads_per_block < threads_per_block_max) {
+    if (batch % 2 == 0 && batch / 2 >= batch_min) {
+      batch /= 2;
+      threads_per_block *= 2;
+    } else if (batch % 3 == 0 && batch / 3 >= batch_min) {
+      batch /= 3;
+      threads_per_block *= 3;
+    } else if (batch % 5 == 0 && batch / 5 >= batch_min) {
+      batch /= 5;
+      threads_per_block *= 5;
+    } else if (batch > batch_max) {
+      threads_per_block = scheduler_utils::roundUpToN(
+          ceilDiv(after_vect, batch_max), dev_prop->warpSize);
+      batch = ceilDiv(after_vect, threads_per_block);
+      break;
+    } else {
+      break;
+    }
+  }
+  // step-2, iteration dim of outer reduction, (inner_dim_numel)
+  // gdimy*unroll*bdimx >= inner_dim_numel
+  int64_t unroll_factor = inner_dim_numel >= 4096 ? 4 : 2;
+  int64_t after_unroll = ceilDiv(inner_dim_numel, unroll_factor);
+  int64_t blocks_per_sm_max1 =
+      ceilDiv(after_unroll, device_multiprocessor_count);
+
+  // this reg_per_thread is just used to estimate threads_per_sm
+  // the actual maxrregcount is set using getRegPerThreadGivenThreadsPerSM
+  // the estimation is based on layer_norm, the estimated result is larger than
+  // actual register usage and no register spill observed
+  constexpr int64_t overhead_register = 40;
+  constexpr int64_t bytes_per_register = 4;
+  const int64_t outer_reduction_partial_result_buffer =
+      batch * vect * sizeof(float) * 2;
+  const int64_t weight_buffer = batch * vect * max_input_dtype_size;
+  const int64_t projected_input_buffer =
+      batch * vect * max_input_dtype_size * 2;
+  const int64_t persistent_buffer_size = outer_reduction_partial_result_buffer +
+      weight_buffer + projected_input_buffer;
+  const int64_t estimated_register_count =
+      persistent_buffer_size / bytes_per_register + overhead_register;
+
+  int64_t reg_per_thread = std::min(estimated_register_count, (int64_t)255);
+  int64_t threads_per_sm = dev_prop->regsPerMultiprocessor / reg_per_thread /
+      dev_prop->warpSize * dev_prop->warpSize;
+  int64_t blocks_per_sm =
+      std::min(blocks_per_sm_max1, threads_per_sm / threads_per_block);
+  int64_t gdimy = blocks_per_sm * device_multiprocessor_count;
+  int64_t bdimx = scheduler_utils::roundUpPow2(ceilDiv(after_unroll, gdimy));
+  int64_t bdimy = threads_per_block / bdimx;
+
+  // if gdimy is too large, reduce it and increase bdimx, decrease bdimy, this
+  // may increase wave tailings
+  const int64_t outer_iter_min = 8;
+  const int64_t gdimy_max = scheduler_utils::roundUpToN(
+      outer_dim_numel / outer_iter_min, device_multiprocessor_count);
+  while (gdimy > gdimy_max && blocks_per_sm > 1) {
+    blocks_per_sm -= 1;
+    gdimy = blocks_per_sm * device_multiprocessor_count;
+    bdimx = scheduler_utils::roundUpToN(ceilDiv(after_unroll, gdimy), 2);
```

Comment:
Don't we need any check about `bdimx`? I assume we don't want to make it too large. Here, we are decreasing `gdimy`, so `bdimx` is increased. It seems we should check if it gets too large.